nifi-extension-bundles/nifi-poi-bundle/nifi-poi-services/src/main/java/org/apache/nifi/excel/StandardCellFieldTypeReader.java

```diff
@@ -0,0 +1,129 @@
+    public DataType getCellDataType(final Cell cell) {
+        Objects.requireNonNull(cell, "Cell required");
+
+        final CellType cellType = cell.getCellType();
+
+        final DataType dataType;
+
+        if (CellType.NUMERIC == cellType) {
+            // Date Formatting check limited to NUMERIC Cell Types
+            if (DateUtil.isCellDateFormatted(cell)) {
+                // Default to TIMESTAMP for date formatted values using standard Date Time Pattern
+                dataType = RecordFieldType.TIMESTAMP.getDataType();
+            } else {
+                final double numericCellValue = cell.getNumericCellValue();
+                final long roundedCellValue = (long) numericCellValue;
+                if (roundedCellValue == numericCellValue) {
+                    dataType = RecordFieldType.LONG.getDataType();
+                } else {
+                    // Default to DOUBLE for NUMERIC values following cell.getNumericCellValue()
+                    dataType = RecordFieldType.DOUBLE.getDataType();
+                }
+            }
+        } else if (CellType.BOOLEAN == cellType) {
+            dataType = RecordFieldType.BOOLEAN.getDataType();
+        } else if (CellType.STRING == cellType) {
+            final String cellValue = cell.getStringCellValue();
+            // Attempt Time Value inference for STRING cell values
+            final Optional<DataType> timeDataType = timeValueInference.getDataType(cellValue);
+            dataType = timeDataType.orElse(RecordFieldType.STRING.getDataType());
+        } else if (CellType.FORMULA == cellType) {
+            dataType = getFormulaResultDataType(cell);
+        } else {
+            // Default to null for known and unknown Cell Types: BLANK, ERROR
+            dataType = null;
+        }
+
+        return dataType;
+    }
+
+    private DataType getFormulaResultDataType(final Cell cell) {
+        final DataType dataType;
+
+        final CellType formulaResultType = cell.getCachedFormulaResultType();
+        if (CellType.BOOLEAN == formulaResultType) {
+            dataType = RecordFieldType.BOOLEAN.getDataType();
+        } else if (CellType.STRING == formulaResultType) {
+            dataType = RecordFieldType.STRING.getDataType();
+        } else if (CellType.NUMERIC == formulaResultType) {
+            // Date Formatting check limited to NUMERIC Cell Types without Conditional Formatting Evaluator
+            if (DateUtil.isCellDateFormatted(cell)) {
+                // Default to TIMESTAMP for date formatted values using standard Date Time Pattern
+                dataType = RecordFieldType.TIMESTAMP.getDataType();
+            } else {
+                // Default to DOUBLE for NUMERIC values following cell.getNumericCellValue()
+                dataType = RecordFieldType.DOUBLE.getDataType();
+            }
+        } else {
+            // Default to null for known and unknown Formula Result Cell Types: BLANK, ERROR
+            dataType = null;
+        }
```

Comment:
I didn't realize null type is an option. Doesn't everything need to have an associated type?
Furthermore it is possible to to have a string value of the error which end users may or may not be interested in.

Reply:
The `null` return is valid, and the Field Type Inference class simply ignores it. In that, case, the inferred type is effectively "unknown", however, the standard behavior of record type inferencing falls back to `STRING`. Since `null` is valid as a return type, this effectively indicates that the type is unknown or unable to be determined by the inferencing process.